File: recursion.py
def generate_number_variations(arr):
    variations = []

for num in arr:
    num_str = str(num)
    num_length = len(num_str)
    visited = [False] * num_length

    def generate_permutations(current_index, current_num):
        nonlocal variations
        if current_index == num_length:
            variations.append(current_num)
            return

        for i in range(num_length):
            if not visited[i]:
                visited[i] = True
                generate_permutations(current_index + 1, current_num * 10 + int(num_str[i]))
                visited[i] = False

    generate_permutations(0, 0)

return variations
Example usage:
arr = [123, 456]
print(generate_number_variations(arr))

File: recursion.js
function generateNumberVariations(arr) {
    const variations = [];


    for (const num of arr) {
      const numStr = num.toString();
      const numLength = numStr.length;
      // console.log(length ${numLength})
      const visited = Array(numLength).fill(false);

  console.log(` visited ${visited}`)
  
  // Use recursion to generate all permutations of the digits
  function generatePermutations(currentIndex, currentNum) {
    if (currentIndex === numLength) {
      variations.push(currentNum);
      // console.log(`variation ${variations}`);
      return;
    }
  
    for (let i = 0; i < numLength; i++) {
      if (!visited[i]) {
        visited[i] = true;
        generatePermutations(currentIndex + 1, currentNum * 10 + parseInt(numStr[i]));
        // console.log(`current num${currentNum}`)
        visited[i] = false;
        // console.log(visited[i])
      }
    }
  }
  
  generatePermutations(0, 0);
  // function filterOut(numStr){
  //   let filtered = numStr

  // }
}
  
return variations;
  }

 const arr= [9041]
 const variations=generateNumberVariations(arr)
 console.log(variations)
